commandapi-core/src/main/java/dev/jorel/commandapi/arguments/CommandArgument.java

```diff
@@ -0,0 +1,203 @@
+
+			SuggestionsBuilder argumentBuilder = new SuggestionsBuilder("", 0);
+			if (hasReplacement(0)) {
+				replacements[0].suggest(new SuggestionInfo(sender, new Object[0], "", ""), argumentBuilder);
+				Suggestions suggestions = argumentBuilder.build();
+				List<String> results = suggestions.getList().stream().map(Suggestion::getText).toList();
+				if (!results.contains(commandLabel))
+					throw CommandSyntaxException.BUILT_IN_EXCEPTIONS.dispatcherUnknownCommand().createWithContext(context);
+			}
+
+			// Enforce argument suggestions
+			String[] arguments = command.substring(command.indexOf(" ") + 1).split(" ");
+			if (!arguments[0].isEmpty() && command.endsWith(" ")) {
+				arguments = Arrays.copyOf(arguments, arguments.length + 1);
+				arguments[arguments.length - 1] = "";
+			}
+
+			List<String> previousArguments = new ArrayList<>();
+			StringBuilder currentInput = new StringBuilder(commandLabel);
+			context.setCursor(currentInput.length());
+
+			for (int i = 0; i < arguments.length - 1; i++) {
+				String currentArgument = arguments[i];
+				argumentBuilder = new SuggestionsBuilder(currentInput.toString(), currentInput.length());
+				if (hasReplacement(i + 1)) {
+					replacements[i + 1].suggest(new SuggestionInfo(sender, previousArguments.toArray(), currentInput.toString(), ""), argumentBuilder);
+					Suggestions suggestions = argumentBuilder.build();
+					List<String> results = suggestions.getList().stream().map(Suggestion::getText).toList();
+					if (!results.contains(currentArgument))
+						throw CommandSyntaxException.BUILT_IN_EXCEPTIONS.dispatcherUnknownArgument().createWithContext(context);
+				}
+				currentInput.append(" ").append(currentArgument);
+				previousArguments.add(currentArgument);
+				context.setCursor(currentInput.length());
+			}
+
+			builder = builder.createOffset(builder.getStart() + command.lastIndexOf(" ") + 1);
+
+			int lastIndex = arguments.length - 1;
+			if (hasReplacement(lastIndex + 1))
+				return replacements[lastIndex + 1].suggest(new SuggestionInfo(sender, Arrays.copyOf(arguments, lastIndex), command, arguments[lastIndex]), builder);
+
+			// Get location sender is looking at if they are a Player, matching vanilla behavior
+			// No builtin Commands use the location parameter, but they could
+			Location location = null;
+			if (sender instanceof Player player) {
+				Block block = player.getTargetBlockExact(5, FluidCollisionMode.NEVER);
+				if (block != null) location = block.getLocation();
+			}
+
+			// Build suggestions for new argument
+			target.tabComplete(sender, commandLabel, arguments, location).forEach(builder::suggest);
+			return builder.buildFuture();
+		});
+	}
+
+	private boolean hasReplacement(int index) {
+		return replacements.length > index && replacements[index] != null;
+	}
+
+	ArgumentSuggestions[] replacements = new ArgumentSuggestions[0];
+
+	public CommandArgument replaceSuggestions(ArgumentSuggestions... suggestions) {
```

Comment:
Needs JavaDocs.

Reply:
The suggestion replacing using `replaceSuggestions(ArgumentSuggestions...)` is cool, but there is one problem with it: It doesn't support _multiple commands_.

If I (for some bizarre reason) wanted to have a command argument that accepts two commands (e.g. `/give [online player] (diamond/minecraft:diamond)` and `/tp [online player] [target player]`), you can't do that with the current implementation (as far as I'm aware).

Reply:
Multiple commands can be supported like this:
```java
new CommandAPICommand("multiplecommands")
        .withArguments(
                new CommandArgument("command").replaceSuggestions(
                        ArgumentSuggestions.strings("give", "tp"),
                        ArgumentSuggestions.strings(info -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)),
                        ArgumentSuggestions.strings(info -> switch ((String) info.previousArgs()[0]) {
                            case "give" -> new String[]{"diamond", "minecraft:diamond"};
                            case "tp" -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new);
                            // This default case should never happen thanks to CommandArgument's suggestion enforcement
                            default -> throw new IllegalStateException("Only give or tp command should have been passed in");
                        })
                )
        ).executes((sender, args) -> {
            ((CommandResult) args[0]).execute(sender);
        }).register();
```
Since the previousArgs given to the SuggestionInfo includes the command label, you can use that to switch between different suggestion paths. I suppose this is an important detail to emphasize: the SuggestionInfo object here holds information for the command being input to the CommandArgument and you cannot access the information from the arguments before the CommandArgument.

Perhaps a more intuitive system could be developed with some sort of tree structure that leads to different ArgumentSuggestions objects, but I think this makes enough sense.

Reply:
Good point - I didn't think of that. It's a bit clunky (which can be mitigated by having suitable documentation which isn't a big issue), but I think having a cleaner implementation that's easier for users may be worth considering. Perhaps something of the form...

```java
replaceSuggestions(SomeClassName...)
```

```java
.replaceSuggestions(
	SomeClassName.of(
		ArgumentSuggestions.strings("give"),
		ArgumentSuggestions.strings(info -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)),
		ArgumentSuggestions.strings("diamond", "minecraft:diamond"),
		ArgumentSuggestions.empty()
	),
	SomeClassName.of(
		ArgumentSuggestions.strings("tp"),
		ArgumentSuggestions.strings(info -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)),
		ArgumentSuggestions.strings(info -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)),
		ArgumentSuggestions.empty()
	)
)
```

It's literally syntactic sugar, but I think it'll make this feature more discoverable and save on dev brain power. This could be a possibility _in addition to_ the existing `replaceSuggestions(ArgumentSuggestions... suggestions)` - I think the method you've already implemented makes a lot of sense.

Reply:
> Needs JavaDocs.

Can confirm implemented.

Reply:
Yeah, another method in CommandArgument like `replaceBranchingSuggesstions` might be nice. I'll try something and see how it goes.